Source/JavaScriptCore/runtime/PropertyTable.h

```diff
@@ -479,59 +512,104 @@ inline PropertyTable* PropertyTable::copy(VM& vm, unsigned newCapacity)
 #ifndef NDEBUG
 inline size_t PropertyTable::sizeInMemory()
 {
-    size_t result = sizeof(PropertyTable) + dataSize();
+    size_t result = sizeof(PropertyTable) + dataSize(isCompact());
     if (m_deletedOffsets)
         result += (m_deletedOffsets->capacity() * sizeof(PropertyOffset));
     return result;
 }
 #endif
 
-inline void PropertyTable::reinsert(const ValueType& entry)
+template<typename Index, typename Entry>
+inline void PropertyTable::reinsert(Index* indexVector, Entry* table, const ValueType& entry)
 {
 #if DUMP_PROPERTYMAP_STATS
     ++propertyTableStats->numReinserts;
 #endif
 
     // Used to insert a value known not to be in the table, and where
     // we know capacity to be available.
-    ASSERT(canInsert());
-    find_iterator iter = find(entry.key);
-    ASSERT(!iter.first);
+    ASSERT(canInsert(entry));
+    auto [emptyIndex, index, offset, attributes] = findImpl(indexVector, table, entry.key());
+    ASSERT_UNUSED(offset, offset == invalidOffset);
+    ASSERT_UNUSED(emptyIndex, emptyIndex == EmptyEntryIndex);
 
     unsigned entryIndex = usedCount() + 1;
-    m_index[iter.second] = entryIndex;
-    table()[entryIndex - 1] = entry;
+    indexVector[index] = entryIndex;
+    table[entryIndex - 1] = entry;
 
     ++m_keyCount;
 }
 
-inline void PropertyTable::rehash(VM& vm, unsigned newCapacity)
+inline void PropertyTable::rehash(VM& vm, unsigned newCapacity, bool forceNonCompact)
 {
 #if DUMP_PROPERTYMAP_STATS
     ++propertyTableStats->numRehashes;
 #endif
 
-    size_t oldDataSize = dataSize();
-    unsigned* oldEntryIndices = m_index;
-    iterator iter = this->begin();
-    iterator end = this->end();
+    bool oldIsCompact = isCompact();
+    size_t oldDataSize = dataSize(oldIsCompact);
+    uintptr_t oldEntryIndices = m_indexVector;
+    unsigned oldIndexSize = m_indexSize;
+    unsigned oldUsedCount = usedCount();
 
     m_indexSize = sizeForCapacity(newCapacity);
     m_indexMask = m_indexSize - 1;
     m_keyCount = 0;
     m_deletedCount = 0;
 
-    m_index = static_cast<unsigned*>(PropertyTableMalloc::zeroedMalloc(dataSize()));
-
-    for (; iter != end; ++iter) {
-        ASSERT(canInsert());
-        reinsert(*iter);
+    // Once table gets non-compact, we do not change it back to compact again.
+    // This is because some of property offset can be larger than UINT8_MAX already.
+    bool isCompact = !forceNonCompact && oldIsCompact && tableCapacity() < UINT8_MAX;
+    size_t newDataSize = dataSize(isCompact);
+    uintptr_t indexVector = bitwise_cast<uintptr_t>(PropertyTableMalloc::zeroedMalloc(newDataSize));
+    m_indexVector = (indexVector | (isCompact ? isCompactFlag : 0));
+
+    if (isCompact) {
```

Comment:
I wonder if these 3 different cases can be abstracted into a lambda that just takes auto* for old and new vector